### DISCUSSION TITLE
Defining External Models

### DISCUSSION THREAD

**Discussion by Coffee2Bits** [DevOps-MBSE/AaC#69] [Archive] [answered]
The tilde was actually intentional, but that doesn't matter right now.  I had used the ~ to indicate that the type referenced in the step wasn't defined in the model.  As long as we can generate Puml, we'll accept this and come up with a more obvious way to handle that later.

_Originally posted by @jondavid-black in https://github.com/jondavid-black/AaC/pull/58#pullrequestreview-786643173_

**Comment by Coffee2Bits** [comment]
The tilde was being transferred to the puml diagram output [example](www.plantuml.com/plantuml/png/RP5DJiGm38NtFeKl01AQNHTe-15BG08EO5eFZTGcLUpCY0rdnuqoYY9bPTdVUyzkSnIbg6bTG9qkZAzXTPuMF2tXWk644ocXpZ6adzage_oAydeBczdSw3RoYdzVmX59C5bjo4TM6oe-s1LuSrzQ770ABxOAIcyaAmQCB5lmmW3GE-tYT7Q-Sr5_Rmivdjt8Rvwpu3pUjJXzIxMNfflmEKU-PXpg9-H87F0zXf55dadf0VQPiHfeWgqA7i50at1e4vuWppg_xiX-iWVy1m00) and it was preventing the diagram from rendering. 

We can put the tilde back and strip it during diagram generation, or we can move this to a discussion about how to handle type references in the DSL that aren't defined in our models because they're outside entities.

**Comment by jondavid-black** [comment]
Probably best handled as a discussion.  I think we'll need a way to represent external entities and their interfaces without fully modeling them.  Maybe we can just do that now by building a super simple representation (i.e. model stub?) of the external entity? Perhaps I was trying to solve a problem that doesn't need to be solved?

**Reply by Coffee2Bits** [reply]
- [x] Converted this issue to a discussion.

**Comment by Coffee2Bits** [comment, marked as answer]
Models currently only require the name and behavior attributes, which does give users a fairly small minimum requirement to define an entity without modeling it completely. That being said, requiring the behavior attribute does introduce enough complexity to warrant a sort of model stub like you suggested, which muddies the problem even more since we use behaviors to define our models' interfaces.

I think we should consider pairing down some of the required fields to allow users to define more minimal models. If we removed the requirement in `Behavior` on `acceptance` then we could allow users to quickly model external entities using the same infrastructure as all of the "completely" modeled entities.

Example 'minimal' model without behavior acceptance:
```
---
model:
    name: ExternalEntity
    behavior:
      - name: SomeAction
        type: request-response
        input: 
          - name: request
            type: string
        output:
          - name: response
             type: string
---
```

I think the above model definition is sufficiently small/simple to avoid needing another definition for external models and/or another definition for simplified interfaces. 

If there isn't an issue removing `acceptance` as a required field in `Behavior`, then this is my preferred solution.

**Reply by jondavid-black** [reply]
Let's do it!

**Comment by lizzcondrey** [comment]
I don't want to close this discussion because I think this is a useful addition to documentation. We came close with our recent addition of [model definitions](https://arch-as-code.org/project_documentation/user_guide/model_definitions.html), and explaining how a paired down model would be seen as only an external entity. But we didn't provide an example of what that looks like. I think that would be a useful add after seeing this discussion thread. Thoughts?
